src/main/java/io/fabric8/elasticsearch/plugin/acl/SearchGuardRoles.java

```diff
@@ -0,0 +1,255 @@
+				
+				public void setType(String type) {
+					this.type = type;
+				}
+				
+				public List<String> getActions() {
+					return actions;
+				}
+				
+				public void setActions(List<String> actions) {
+					this.actions = actions;
+				}
+				
+				@Override
+				public String toString() {
+					return ToStringBuilder.reflectionToString(this, ToStringStyle.MULTI_LINE_STYLE);
+				}
+			}
+			
+			@Override
+			public String toString() {
+				return ToStringBuilder.reflectionToString(this, ToStringStyle.MULTI_LINE_STYLE);
+			}
+		}
+	}
+
+	@Override
+	public Iterator<Roles> iterator() {
+		return new ArrayList<>(roles).iterator();
+	}
+	
+	@Override
+	public String toString() {
+		return ToStringBuilder.reflectionToString(this, ToStringStyle.MULTI_LINE_STYLE);
+	}
+	
+	public void removeRole(Roles role) {
+		roles.remove(role);
+	}
+	
+	public void syncFrom(UserProjectCache cache, final String userProfilePrefix){
+		removeSyncAcls();
+		
+		RolesBuilder builder = new RolesBuilder();
+		
+		for (String project : cache.getAllProjects() ) {
+			String projectName = String.format("%s_%s", PROJECT_PREFIX, project.replace('.', '_')); 
+			String indexName = String.format("%s?*", project.replace('.', '?'));
+			
+			RoleBuilder role = new RoleBuilder()
+					.setName(projectName)
+					.setActions(indexName, DEFAULT_ROLE_TYPE, DEFAULT_ROLE_ACTIONS);
+			
+			builder.addRole(role.build());
+		}
+		
+		for (Map.Entry<String, Set<String>> userProjects : cache.getUserProjects().entrySet()) {
+			String usernameHash = getUsernameHash(userProjects.getKey());
+			String projectName = String.format("%s_%s_%s", ROLE_PREFIX, "kibana", usernameHash); 
+			String indexName = String.format("%s?%s", userProfilePrefix.replace('.', '?'), usernameHash);
+			
+			RoleBuilder role = new RoleBuilder()
+					.setName(projectName)
+					.setActions(indexName, DEFAULT_ROLE_TYPE, KIBANA_INDEX_ACTIONS)
+					.setActions(DEFAULT_ROLE_INDEX, DEFAULT_ROLE_TYPE, DEFAULT_ROLE_ACTIONS)
+					.setCluster(DEFAULT_CLUSTER_ACTIONS);
+			
+			builder.addRole(role.build());
+		}
+		
+		roles.addAll(builder.build());
+	}
+	
+	// Remove roles that start with "gen_"
+	private void removeSyncAcls() {
+		for (Roles role : new ArrayList<>(roles)) {
+			if(role.getName() != null && role.getName().startsWith(ROLE_PREFIX)){
+				removeRole(role);
+			}
+		}
+	}
+	
+	@SuppressWarnings("unchecked")
+	public SearchGuardRoles load(Map<String, Object> source) {
+		
+		RolesBuilder builder = new RolesBuilder();
+		
+		for ( String key : source.keySet() ) {
+			RoleBuilder roleBuilder = new RoleBuilder().setName(key);
+			
+			// get out cluster and indices
+			HashMap<String, Object> role = (HashMap<String, Object>) source.get(key);
+			
+			List<String> cluster = (ArrayList<String>)role.get(CLUSTER_HEADER);
+			roleBuilder.setCluster(cluster);
+			
+			HashMap<String, HashMap<String, ArrayList<String>>> indices = (HashMap<String, HashMap<String, ArrayList<String>>>) role.get(INDICES_HEADER);
```

Comment:
Any change this can be modified while processing thus throwing a modification exception?